src/cpuinfo_x86.c

```diff
@@ -1082,27 +1097,106 @@ static void ParseLeaf4(const int max_cpuid_leaf, CacheInfo* info) {
 // Internal structure to hold the OS support for vector operations.
 // Avoid to recompute them since each call to cpuid is ~100 cycles.
 typedef struct {
-  bool have_sse;
+  bool have_sse_via_os;
+  bool have_sse_via_cpuid;
   bool have_avx;
   bool have_avx512;
   bool have_amx;
 } OsSupport;
 
+static const OsSupport kEmptyOsSupport;
+
+static OsSupport CheckOsSupport(const uint32_t max_cpuid_leaf) {
+  const Leaf leaf_1 = SafeCpuId(max_cpuid_leaf, 1);
+  const bool have_xsave = IsBitSet(leaf_1.ecx, 26);
+  const bool have_osxsave = IsBitSet(leaf_1.ecx, 27);
+  const bool have_xcr0 = have_xsave && have_osxsave;
+
+  OsSupport os_support = kEmptyOsSupport;
+
+  if (have_xcr0) {
+    // AVX capable cpu will expose XCR0.
+    const uint32_t xcr0_eax = GetXCR0Eax();
+    os_support.have_sse_via_cpuid = HasXmmOsXSave(xcr0_eax);
+    os_support.have_avx = HasYmmOsXSave(xcr0_eax);
+    os_support.have_avx512 = HasZmmOsXSave(xcr0_eax);
+    os_support.have_amx = HasTmmOsXSave(xcr0_eax);
+  } else {
+    // Atom based or older cpus need to ask the OS for sse support.
+    os_support.have_sse_via_os = true;
+  }
+
+  return os_support;
+}
+
+#if defined(HAVE_SYSCTLBYNAME)
+static bool SysCtlByName(const char* name) {
+  int enabled;
+  size_t enabled_len = sizeof(enabled);
+  const int failure = sysctlbyname(name, &enabled, &enabled_len, NULL, 0);
+  return failure ? false : enabled;
+}
+#endif
+
+static void DetectSseViaOs(X86Features* features) {
+#if defined(CPU_FEATURES_COMPILER_MSC)
+  // https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-isprocessorfeaturepresent
+  features->sse = IsProcessorFeaturePresent(PF_XMMI_INSTRUCTIONS_AVAILABLE);
+  features->sse2 = IsProcessorFeaturePresent(PF_XMMI64_INSTRUCTIONS_AVAILABLE);
+  features->sse3 = IsProcessorFeaturePresent(PF_SSE3_INSTRUCTIONS_AVAILABLE);
+#elif defined(HAVE_UTSNAME_H)
+  struct utsname buf;
+  uname(&buf);
+  if (CpuFeatures_StringView_IsEquals(str(buf.sysname), str("Darwin"))) {
+#if defined(HAVE_SYSCTLBYNAME)
+    // Handling Darwin platform through sysctlbyname when available.
+    features->sse = SysCtlByName("hw.optional.sse");
+    features->sse2 = SysCtlByName("hw.optional.sse2");
+    features->sse3 = SysCtlByName("hw.optional.sse3");
+    features->ssse3 = SysCtlByName("hw.optional.supplementalsse3");
+    features->sse4_1 = SysCtlByName("hw.optional.sse4_1");
+    features->sse4_2 = SysCtlByName("hw.optional.sse4_2");
```

Comment:
so on Apple for fake_cpu you also need to moc SysCtlByName...

Reply:
Yes I was expecting this :)
Working on a fix.